media/windows/ECC-MQTT-IoT/ECC_MQTT_IoT_SQLite.py

```diff
@@ -101,16 +124,149 @@
     print(F"ECC MQTT IoT Listener application, version {eccmqtt_iot_version}, {eccmqtt_iot_date}...")
     sys.exit(0)
 
+
+class GZipRotator(logging.handlers.RotatingFileHandler):
+
+    def doRollover(self):
+
+        global done_flag
+
+        # Let the default log rollover routine do its thing first...
+        super().doRollover()
+
+        # Construct a file pattern to match all current gzip'd logs
+        base_file_pattern = self.baseFilename + ".[0-9].gz"
+        # Get a list of the current .gzip'd log files to determine the next extension to use
+        arch_files = glob.iglob(base_file_pattern)
+        # Loop through the archived log files to find the last file number in use
+        lastgz_file_num = '0'  # default value for first iteration (no current gzips)
+        for archFile in arch_files:
+            fn = archFile.split('.')
+            # The gzip'd file "number" is in array element 2
+            lastgz_file_num = fn[2]
+        lastgz_file_int = int(lastgz_file_num)
+        # Construct a gzip file name for the destination by incrementing the last file number
+        # used.  If more than (backupCount) files are already zipped, delete the oldest and rename
+        # the remaining to free up the last number for re-use.  This ensures that the newest gzip
+        # will always be the highest number.
+        nextgz_file_int = lastgz_file_int + 1
+        if nextgz_file_int > self.backupCount:
+            nextgz_file_int = self.backupCount
+            # Since we already have the maximum gzip'd logs, delete the oldest, then rename the
+            # remaining to free up the last slot.
+            oldest_gz_log = self.baseFilename + ".1.gz"
+            os.remove(oldest_gz_log)
+            # Rename the remaining gzip'd logs...
+            arch_files = glob.iglob(base_file_pattern)
+            for archFile in arch_files:
+                fn = archFile.split('.')
+                # The gzip'd file "number" is in array element 2
+                newgz_file_num = int(fn[2]) - 1
+                newgz_file = fn[0] + "." + fn[1] + "." + str(newgz_file_num) + ".gz"
+                os.rename(archFile, newgz_file)
+        else:
+            # haven't hit the maximum backup file count yet, so simply construct the
+            # next gzip file name based on the existing base file name components
+            fn = (self.baseFilename + ".1.gz").split('.')
+
+        nextgz_file_name = fn[0] + "." + fn[1] + "." + str(nextgz_file_int) + ".gz"
+        dest = self.baseFilename + ".1"
```

Comment:
If you're rotating, I think you always need to rename N files.  It might be simpler to do something like this inseatd of your iglob scheme (typed in Github / not tested):

```python
# This will iterate from foo.(BACKUPCOUNT-1).gz to foo.1.gz
for i in range(self.backupCount - 1, 0, -1):
    filename = f'{self.baseFilename}.{i}.gz'
    if os.path.exists(filename):
        next_filename = f'{self.baseFilename}.{i+1}.gz'
        os.rename(filename, next_filename)

# Now rename the "zeroth" file ("foo") to foo.1.
next_filename = f'{self.baseFilename}.1'
os.rename(self.baseFilename, next_filename)

# ...now you can advance on to:
# 1. Fire off the thread to gzip foo.1 to foo.1.gz
# 2. Write the log message to a new foo file (which I assume is handled by the parent class?)
```

Does that make sense, or did I mis-understand the scheme?

Reply:
If you're rotating, and the current number of backup files is < the max specified, then renaming isn't necessary.  This only applies in the case where there are already the max backups specified, and we're rotating the oldest "off the stack" and renaming the others.  (The oldest archive is ".1", then ".2", etc.)  If I understand your code, it looks like you've reversed this, such that the newest archive is the lowest numbered, but maybe I've missed something.

Reply:
Ah, yes, perhaps this is the disconnect.  The rotating logs I'm used to are exactly as you described:

1. `foo`: the logfile that we're currently writing to, from time "now" back to time A
1. `foo.1[.gz]`: the log from time A back to time B
1. `foo.2[.gz]`: the log from time B back to time C
1. `foo.3[.gz]`: the log from time C back to time D
1. ...etc.

I.e., as you surmised:

- the "zeroth" file is the current file
- .1 is the next-oldest
- .N is the oldest

That's how the Linux logrotater works.  From my understanding, this is how the Python RotatingFileHandler works, too: https://docs.python.org/3/library/logging.handlers.html#rotatingfilehandler

Should we change that convention and go the other way?

Reply:
Is there an advantage you see in changing the convention that I'm missing?  I would suggest leaving it as the default handler works, since that's what I'm used to seeing, and changing it to reverse the order might be more confusing down the road.  Just my 2 cents though...

Reply:
Yeah, I think we're in another toe-MAY-toe vs. toe-MAH-toe situation.  There's probably no strong technical reason to go one way vs. the other.

I think there is a _small_ procedural reason to lean towards the "larger number means older file" convention: the other Linux and Python loggers use that convention.  I.e., this would have the Gzip'ed behave the same as the rest of the ecosystem.

FWIW, the other ECC Linux scripts have been using the Linux logrotator and Python file rotators for a while.